File: python/bank-account/bank_account.py
import threading

class BankAccount:
    def __init__(self):
        self.isOpen = False
        self.balance = 0

    def get_balance(self):
        if(not self.isOpen):
            raise ValueError("This account is closed")

        return self.balance

    def open(self):
        if(self.isOpen):
            raise ValueError("This account is already opened")

        self.isOpen = True
        self.balance = 0

    def deposit(self, amount):
        if(not self.isOpen):
            raise ValueError("This account is closed")

        if(amount < 0):
            raise ValueError("You can't deposit a negative amount")

        threadLock.acquire()
        self.balance += amount
        threadLock.release()

    def withdraw(self, amount):
        if(amount > self.balance):
            raise ValueError("You don't have enough money to do that")

        if(amount < 0):
            raise ValueError("You can't withdraw a negative amount")

        threadLock.acquire()
        self.balance -= amount
        threadLock.release()

    def close(self):
        if(not self.isOpen):
            raise ValueError("This account is already closed")

        self.isOpen = False

threadLock = threading.Lock()

File: javascript/bank-account/bank-account.js


export class BankAccount {
  constructor() {
    this.isOpen = false;
    this.accountBalance = 0;
  }

  open() {
    if(this.isOpen){
      throw new ValueError();
    }

    this.isOpen = true;
    this.accountBalance = 0;

  }

  close() {
    if(!this.isOpen){
      throw new ValueError();
    }

    this.isOpen = false;
  }

  deposit(depositAmount) {
    if(!this.isOpen){
      throw new ValueError();
    }

    // can't be negative
    if(depositAmount < 0){
      throw new ValueError();
    }


    this.accountBalance += depositAmount;
  }

  withdraw(withdrawAmount) {
    // can't withdraw on a closed account
    if(!this.isOpen){
      throw new ValueError();
    }

    // can't withdraw more than what you have
    if(withdrawAmount > this.accountBalance){
      throw new ValueError();
    }

    // can't withdraw negative money
    if(withdrawAmount < 0){
      throw new ValueError();
    }

    this.accountBalance -= withdrawAmount;
  }

  get balance() {
    if(!this.isOpen){
      throw new ValueError();
    }

    return this.accountBalance;
  }

  set balance(newValue){
    this.accountBalance = newValue;
  }

}

export class ValueError extends Error {
  constructor() {
    super('Bank account error');
  }
}
